src/H5FDint.c

```diff
@@ -260,152 +290,2049 @@ H5FD_write(H5FD_t *file, H5FD_mem_t type, haddr_t addr, size_t size, const void
+
+    *vector_was_sorted = TRUE;
+
+    /* if count <= 1, vector is sorted by definition */
+    if (count > 1) {
+
+        /* scan the addrs array to see if it is sorted */
+        i = 1;
+
+        while ((*vector_was_sorted) && (i < (int)(count - 1))) {
+
+            if (H5F_addr_gt(addrs[i - 1], addrs[i])) {
+
+                *vector_was_sorted = FALSE;
+            }
+            else if (H5F_addr_eq(addrs[i - 1], addrs[i])) {
+
+                HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "duplicate addr in vector")
+            }
+            i++;
+        }
+    }
+
+    if (*vector_was_sorted) {
+
+        *s_types_ptr = types;
+        *s_addrs_ptr = addrs;
+        *s_sizes_ptr = sizes;
+        *s_bufs_ptr  = bufs;
+    }
+    else {
+
+        /* must sort the addrs array in increasing addr order, while
+         * maintaining the association between each addr, and the
+         * sizes[], types[], and bufs[] values at the same index.
+         *
+         * Do this by allocating an array of struct H5FD_vsrt_tmp_t, where
+         * each instance of H5FD_vsrt_tmp_t has two fields, addr and index.
+         * Load the array with the contents of the addrs array and
+         * the index of the associated entry. Sort the array, allocate
+         * the s_types_ptr, s_addrs_ptr, s_sizes_ptr, and s_bufs_ptr
+         * arrays and populate them using the mapping provided by
+         * the sorted array of H5FD_vsrt_tmp_t.
+         */
+        int    j;
+        int    fixed_size_index = (int)count;
+        int    fixed_type_index = (int)count;
+        size_t srt_tmp_size;
+
+        srt_tmp_size = ((size_t)count * sizeof(struct H5FD_vsrt_tmp_t));
+
+        if (NULL == (srt_tmp = (H5FD_vsrt_tmp_t *)HDmalloc(srt_tmp_size)))
+
+            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "can't alloc srt_tmp")
+
+        for (i = 0; i < (int)count; i++) {
+
+            srt_tmp[i].addr  = addrs[i];
+            srt_tmp[i].index = i;
+        }
+
+        /* sort the srt_tmp array */
+        HDqsort(srt_tmp, (size_t)count, sizeof(struct H5FD_vsrt_tmp_t), H5FD__vsrt_tmp_cmp);
+
+        /* verify no duplicate entries */
+        i = 1;
+
+        while (i < (int)(count - 1)) {
+
+            HDassert(H5F_addr_lt(srt_tmp[i - 1].addr, srt_tmp[i].addr));
+
+            if (H5F_addr_eq(addrs[i - 1], addrs[i])) {
+
+                HGOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "duplicate addr in vector")
+            }
+            i++;
+        }
+
+        if ((NULL == (*s_types_ptr = (H5FD_mem_t *)HDmalloc((size_t)count * sizeof(H5FD_mem_t)))) ||
+            (NULL == (*s_addrs_ptr = (haddr_t *)HDmalloc((size_t)count * sizeof(haddr_t)))) ||
+            (NULL == (*s_sizes_ptr = (size_t *)HDmalloc((size_t)count * sizeof(size_t)))) ||
+            (NULL == (*s_bufs_ptr = (void *)HDmalloc((size_t)count * sizeof(void *))))) {
+
+            HGOTO_ERROR(H5E_RESOURCE, H5E_CANTALLOC, FAIL, "can't alloc sorted vector(s)")
+        }
```

Comment:
Why not sort the arrays in place?

Reply:
Because those arrays are owned by a different section of the code, and potentially could be owned by the application. We could potentially add an internal option to allow this code to reorder those arrays. I don't have time for that now though.